### PR TITLE
add_message_size_read

### LINKED ISSUE #4: Define configs for each transports

### LINKED ISSUE #37: implement a max message size to read

### PR DESCRIPTION
**Pull request by modship**
close #37 

maybe close #4 too

### CODE REVIEW

File: src/transports/tcp.rs [hunk 1]
```diff
@@ -414,24 +421,28 @@ impl<Id: PeerId> Transport<Id> for TcpTransport<Id> {
...
+        config: &Self::TransportConfig,
+    ) -> PeerNetResult<Vec<u8>> {
         //TODO: Config one
         let mut len_bytes = vec![0u8; 4];
         endpoint
             .stream
             .read_exact(&mut len_bytes)
             .map_err(|err| TcpError::ConnectionError.wrap().new("recv len", err, None))?;
-        let res_size = u32::from_be_bytes(len_bytes.try_into().map_err(|err| {
+        let res_size = usize::from_be_bytes(len_bytes.try_into().map_err(|err| {
```
Comment (AurelienFT) [inline src/transports/tcp.rs:434]: Can you change this in the `send` also ? :)

File: tests/limits.rs [hunk 2]
```diff
@@ -41,6 +41,7 @@ fn check_multiple_connection_refused() {
         init_connection_handler: DefaultInitConnection {},
         optional_features: PeerNetFeatures::default(),
         message_handler: DefaultMessagesHandler {},
+        max_message_size_read: 1048576000,
```
Comment (litchipi) [inline tests/limits.rs:44]: Could there be tests that have a very low `max_message_size_read` to check it limits well ?
This new feature doesn't seam tested now

File: src/transports/tcp.rs (hunk 1 above)
Comment (litchipi) [reply src/transports/tcp.rs:434]: I'm not sure having `usize` is good when exchanging data over the network as it's not fixed size across architectures.
Having a `u32` or `u64` would be better there IMO

File: src/transports/tcp.rs [hunk 3]
```diff
@@ -311,11 +316,12 @@ impl<Id: PeerId> Transport<Id> for TcpTransport<Id> {
         context: Ctx,
         address: SocketAddr,
         timeout: Duration,
-        _config: &Self::OutConnectionConfig,
+        _config: &Self::TransportConfig,
         message_handler: M,
         handshake_handler: I,
     ) -> PeerNetResult<JoinHandle<PeerNetResult<()>>> {
         let peer_stop_rx = self.peer_stop_rx.clone();
+        let conf = self.config.clone();
```
Comment (litchipi) [inline src/transports/tcp.rs:324]: Is it on purpose that you use `self.config` and ignore the config passed as parameters ?
It feels unatural when calling a function with a `config` field that this config is ignored and a struct-tied value is used instead ...

**Review by litchipi — CHANGES_REQUESTED** [review]

File: src/transports/tcp.rs (hunk 1 above)
Comment (modship) [reply src/transports/tcp.rs]: switched to u32

File: src/transports/endpoint.rs [hunk 4]
```diff
@@ -35,10 +36,24 @@ impl Endpoint {
             Endpoint::Quic(endpoint) => QuicTransport::<Id>::send(endpoint, data),
         }
     }
-    pub fn receive<Id: PeerId>(&mut self) -> PeerNetResult<Vec<u8>> {
+    pub fn receive<Id: PeerId>(&mut self, config: ConnectionConfig) -> PeerNetResult<Vec<u8>> {
         match self {
-            Endpoint::Tcp(endpoint) => TcpTransport::<Id>::receive(endpoint),
-            Endpoint::Quic(endpoint) => QuicTransport::<Id>::receive(endpoint),
+            Endpoint::Tcp(endpoint) => {
```
[the remark this answers is not in the document]
Comment (modship) [reply src/transports/endpoint.rs]: done

File: src/transports/mod.rs [hunk 5]
```diff
@@ -116,13 +130,13 @@ impl<Id: PeerId> Transport<Id> for InternalTransportType<Id> {
         context: Ctx,
         address: SocketAddr,
         timeout: Duration,
-        config: &Self::OutConnectionConfig,
+        config: &Self::TransportConfig,
         message_handler: M,
         init_connection_handler: I,
     ) -> PeerNetResult<JoinHandle<PeerNetResult<()>>> {
         match self {
```
[the remark this answers is not in the document]
Comment (modship) [reply src/transports/mod.rs]: done

File: src/transports/mod.rs [hunk 6]
```diff
@@ -160,10 +174,24 @@ impl<Id: PeerId> Transport<Id> for InternalTransportType<Id> {
         }
     }
 
-    fn receive(endpoint: &mut Self::Endpoint) -> PeerNetResult<Vec<u8>> {
+    fn receive(endpoint: &mut Self::Endpoint, config: &ConnectionConfig) -> PeerNetResult<Vec<u8>> {
         match endpoint {
```
[the remark this answers is not in the document]
Comment (modship) [reply src/transports/mod.rs]: done

File: tests/limits.rs (hunk 2 above)
Comment (modship) [reply tests/limits.rs:47]: done in limits.rs

File: src/transports/tcp.rs (hunk 3 above)
Comment (modship) [reply src/transports/tcp.rs]: good catch

File: src/transports/mod.rs [hunk 7]
```diff
@@ -116,33 +129,30 @@ impl<Id: PeerId> Transport<Id> for InternalTransportType<Id> {
...
                     context,
                     address,
                     timeout,
                     config,
                     message_handler,
                     init_connection_handler,
                 ),
-                _ => Err(PeerNetError::WrongConfigType.error("try_connect match quic", None)),
-            },
+            _ => Err(PeerNetError::WrongConfigType.error("try_connect match quic", None)),
```
Comment (litchipi) [inline src/transports/mod.rs:155]: It's not specific to quic anymore, so you should maybe change the error message

File: src/transports/tcp.rs [hunk 8]
```diff
@@ -426,7 +436,11 @@ impl<Id: PeerId> Transport<Id> for TcpTransport<Id> {
                 .wrap()
                 .error("recv len", Some(format!("{:?}", err)))
         })?);
-        if res_size > 1048576000 {
+
```
Comment (litchipi) [inline src/transports/tcp.rs:439]: Some debug messages left

**Review by litchipi — APPROVED** [review]

**Pull request merged**